### PR TITLE
[new-rule] No self assign

### PR DESCRIPTION
**Pull request by Zzzen**
#### PR checklist

- [x] Addresses an existing issue: #3976 
- [x] New feature, bugfix, or enhancement
  - [x] Includes tests
- [ ] Documentation update

### CODE REVIEW

**Comment by michaeljota** [comment]
Please see #4141 before you contribute to this project. Thank you!

**Comment by Zzzen** [comment]
ping?

File: src/rules/noSelfAssignRule.ts [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+interface Option {
+    props: boolean;
+}
+
+export class Rule extends Lint.Rules.AbstractRule {
+    /* tslint:disable:object-literal-sort-keys */
+    public static metadata: Lint.IRuleMetadata = {
+        ruleName: "no-self-assign",
+        description: "This rule is aimed at eliminating self assignments.",
```
Comment (JoshuaKGoldberg) [inline src/rules/noSelfAssignRule.ts]: This description doesn't add anything. "This rule is aimed at" is implied; "eliminating" is a longer form of "no"; "assignments" is abbreviated by "assign". Amusing that that's what the ESLint docs call it, but can we get a better description here? Maybe:

> Stops variables from being assigned to themselves.

File: src/rules/noSelfAssignRule.ts [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+                },
+            },
+        },
+        optionExamples: [[true, { props: false }]],
+        type: "maintainability",
+        typescriptOnly: false,
+    };
+
+    public static FAILURE_STRING(name: string) {
+        return `variable '${name}' is assigned to itself.`;
```
Comment (JoshuaKGoldberg) [inline src/rules/noSelfAssignRule.ts]: This isn't always for variables. Either change it to say the type (variable, property, etc.) or just remove that part of the string.

File: src/rules/noSelfAssignRule.ts [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+
+    if (isPropertyAccessExpression(leftExp) && isPropertyAccessExpression(rightExp)) {
+        return isSameMember(leftExp, rightExp);
+    }
+
+    return isIdentifier(leftExp) && isIdentifier(rightExp) && leftExp.text === rightExp.text;
+}
+
+function isSameProperty(left: ts.Node, right: ts.Node) {
+    if (isPropertyAccessExpression(left) && isPropertyAccessExpression(right)
```
Comment (JoshuaKGoldberg) [inline src/rules/noSelfAssignRule.ts]: Nit: just directly return the result of this expression. `return isPropertyAccessExpression(left) && ...`

File: test/rules/no-self-assign/default/test.ts.lint [hunk 4]
```diff
@@ -0,0 +1,65 @@
+var a = a;
```
Comment (JoshuaKGoldberg) [inline test/rules/no-self-assign/default/test.ts.lint]: Could you please add some cases with three variables in array and object spreads? Seems like that could be a cause of bugs.

Comment (JoshuaKGoldberg) [inline test/rules/no-self-assign/default/test.ts.lint]: How about some test cases for assignments within operators?

```typescript
a = a, a = b, b = b;
a = a && a == a && b = a || c = a, a = b, b = b;
```

**Comment by Zzzen** [comment]
Since this rule exists in eslint and now tslint is deprecated. I guess this pr should be closed.

**Pull request closed without merging**